### ISSUE TITLE
dotnet build and analyzers

### ISSUE THREAD

**Issue by livarcocc** [dotnet/roslyn-analyzers#1236]
_From @livarcocc on June 15, 2017 5:8_

_From @ppekrol on June 13, 2017 13:36_

Hi,

## Steps to reproduce

I have following project with analyzers that target netstandard1.3
https://github.com/ravendb/raven.codeanalysis

The issue I have is that when using `dotnet build` I'm getting warning as follows:

```
CSC : warning CS8032: An instance of analyzer Raven.CodeAnalysis.ExceptionBlock.TodoCommentOnExceptionHandlerAnalyzer cannot be created from C:\Users\ppekr\.nuget\packages\raven.codeanalysis\1.0.3\analyzers\dotnet\cs\Raven.CodeAnalysis.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=2.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. The system cannot find the file specified.. [C:\workspaces\AnalyzerIssue\AnalyzerIssue\AnalyzerIssue.csproj]
```

 but from VS2017 everything is working as expected.

I've created a sample project that demonstrates the issue
https://github.com/ppekrol/AnalyzerIssue

Just do:
`dotnet restore`
`dotnet build`

I must be doing something terribly wrong or something is not working as expected.

I've tried adding
```
<PackageReference Include="Microsoft.CodeAnalysis.Common" Version="2.2.0" />
<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.2.0" />
```

without any luck.

## Expected  behavior

No warnings

## Environment data
`dotnet --info` output:
```
.NET Command Line Tools (1.0.4)

Product Information:
 Version:            1.0.4
 Commit SHA-1 hash:  af1e6684fd

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.4
```

_Copied from original issue: dotnet/cli#6850_

_Copied from original issue: dotnet/roslyn#20233_

**Comment by livarcocc** [comment]
@tmat can you take a look here?

**Comment by livarcocc** [comment]
_From @tmat on June 13, 2017 23:45_

@mavasani ?

**Comment by livarcocc** [comment]
Let me know if this is the wrong repo for this or if this is a CLI/SDK issue, in which case, please, move this back to the CLI.

**Comment by Pilchie** [comment]
@livarcocc general problems with analyzer infrastructure tend to go right in dotnet/roslyn.

**Comment by mavasani** [comment]
This seems to be an issue that the analyzer package is built against version 2.2.0 of MS.CA, but the analyzer is being run against a different version (most likely lower version) of Microsoft.CodeAnalysis.dll. Which version of Microsoft.CodeAnalysis.dll is being used by your compiler toolset?

**Comment by mavasani** [comment]
I took a quick look at the analyzer source code: https://github.com/ravendb/raven.codeanalysis/blob/master/Raven.CodeAnalysis/ExceptionBlock/TodoCommentOnExceptionHandlerAnalyzer.cs

There seem to be no requirement to be build the analyzer against version 2.2.0 of MS.CA. I would change the analyzer to build with MS.CA version 1.0.0, so that it runs fine against all compiler toolsets.

**Comment by mavasani** [comment]
I am closing this bug as by design, please let us know if changing the MS.CA reference in analyzer to 1.0.0 or running the existing analyzer against compiler toolset >= 2.2.0 doesn't fix your issue.

**Comment by sharwell** [comment]
@mavasani Is CS8032 only reported for a mismatch in the MS.CA assembly version, or for any failure to load the target assembly?

**Comment by mavasani** [comment]
@sharwell - This is a generic diagnostic for analyzer load failure, we include the underlying exception message in the diagnostic message. See http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerFileReference.cs,e99e1eb9b3f1fa5f,references.

**Comment by sharwell** [comment]
@mavasani Thanks, that is a problem. I filed dotnet/roslyn#20316 to get that fixed.

**Comment by ppekrol** [comment]
@mavasani Not sure what version compiler toolset is using. How can I check that? How can I force it to use different one?

I've switched to 1.0.0 and it works fine. Commit for reference https://github.com/ravendb/raven.codeanalysis/commit/c5e404ce3370c48d21823a4ae5af11181586f1fd

**Comment by mavasani** [comment]
> Not sure what version compiler toolset is using. How can I check that?

You can find Microsoft.CodeAnalysis.dll next to your csc.exe in compiler toolset. When you run a build, you should see csc.exe or VBCSCompiler.exe execute in the task manager, and you can open its file location.

You can force your project to use a specific compiler toolset by adding a NuGet package reference to a specific version of https://www.nuget.org/packages/Microsoft.Net.Compilers. However, I am not sure if this package works for netstandard projects. Adding @agocke @jaredpar for guidance.

**Comment by agocke** [comment]
@mavasani The compiler doesn't care what project you're targeting, but the Microsoft.NET.Compilers package only works with the Desktop .NET framework, so it only runs on Windows using MSBuild, not `dotnet build`.

**Comment by ppekrol** [comment]
@mavasani I do not see any `csc.exe` or `VBCSCompiler.exe` during `dotnet build` only `dotnet.exe`
but in 
`C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.5`
and
`C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.1.2`
there are `Microsoft.CodeAnalysis` DLLs with version `1.3.0.60613`, so I assume that this is the max that custom analyzers can use?

**Comment by mavasani** [comment]
>  so I assume that this is the max that custom analyzers can use?

Yes, that is correct. Unless you move to a toolset that contains a higher version of Microsoft.CodeAnalysis.dll. I'd stick with your current PR of building against lower version MS.CA.

**Comment by ppekrol** [comment]
Yes, it works fine now. Thanks for help!

BTW, the version for `Microsoft.CodeAnalysis`  in `C:\Program Files\dotnet\sdk\1.0.4` is `2.0.0-rc5-61427-04`.  Not sure which one compiler is using... Maybe this is the limit... Anyway thanks for help :)

**Comment by agocke** [comment]
@ppekrol The one in the SDK should be the one used during `build`. The others are for when you build applications to target a shared framework, like when you target NETCoreApp1.1.

**Comment by ppekrol** [comment]
@agocke Thanks for explanation.

**Comment by jskeet** [comment]
@agocke wrote:

> At the moment we do not have a method of overriding the compiler in the dotnet CLI. The only way to upgrade the compiler is to install a newer version of CLI.

Is this being tracked in another issue? I really want to be able to start using analyzers in my Travis build if possible...

**Comment by agocke** [comment]
@jskeet I'm using this issue to track https://github.com/dotnet/roslyn/issues/16717

I'd like to know what's blocking you, though. Is pulling down a zip of a newer CLI prohibitive? I've also set our build to upload nightlies of the mentioned `Microsoft.NETCore.Compilers` package on our MyGet feed, but that's obviously pre-release level. Is it just that you want an official release build?

**Comment by jskeet** [comment]
@agocke: Thanks for the bug link.

In terms of downloading nightlies etc: I have a few problems with that:

- Using a whole CLI is a bigger system change than just adding a reference to a package. If I'm testing a non-GA C# compiler, I want to be really, really sure that doesn't affect anything other than the project I'm installing it in.
- I may be trying to use released C# compiler bits (e.g. Roslyn 2.2) but without there being a CLI version which has gone GA in parallel. For production builds I want to use only production bits, including the CLI.
- It's significantly easier to add a package reference than to modify multiple CI scripts to install a specific CLI version, when they've already got the GA CLI installed
- While it's reasonable for one person to go through a bit of pain to get an experimental build going, I don't want to have to give instructions for the whole team (and any other potential contributor) to download a different CLI in order to work on the codebase.

Now since posting yesterday, I've found a different solution to my problem. I had some Roslyn 1.x analyzers that I wanted to use in our CLI build on Linux, so I wanted Roslyn 2.x. Adding the analyzer caused an error of the 2.2 Roslyn DLLs (latest GA) not being found. I've downgraded to 2.0 and everything is working fine. So my immediate problem is fixed, but I'm still very interested in a longer term solution. (Partly this is due to writing about future versions of C# in a personal capacity too... I'd really like to be able to do that just by adding a package reference.)

**Comment by agocke** [comment]
@jskeet Very helpful, thanks. Regarding your analyzers, were they actually using APIs that were only added in 2.0+? 

I ask because I'm worried that some people are needlessly upgrading analyzer references because there are "newer" CodeAnalysis binaries, even though analyzers always run on the "user" compiler. That is, the authors don't realize that analyzers will always run on the compiler the user is running, not what they built against.

**Comment by jskeet** [comment]
@agocke: No, the analyzers were written against Roslyn 1.x, but we want to be able to run them on Travis on Linux, so we need the portable versions. As I understand it, Roslyn 1.x only targets the desktop framework.

**Comment by sharwell** [comment]
@jskeet StyleCop Analyzers targets Roslyn 1.x and is being used in Linux builds: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/2351#issuecomment-305529327

**Comment by jskeet** [comment]
@sharwell: Hmm... I'm now confused as to why I thought it wouldn't work before. Looks like we could indeed have stuck with 1.3.2.

I'd still like to be able to use different compiler versions from dotnet build without installing a new SDK, mind you :)

**Comment by jskeet** [comment]
Hmm... have just tried using `dotnet build` with a Roslyn 3.0 prelease version of `Microsoft.Net.Compilers` and it seems to work fine (in terms of picking up C# 7.1 features). So it looks like it works for *some* aspects, but not necessarily all.
